storagestate/storagestate.go

```diff
@@ -0,0 +1,709 @@
+	isSamePartition     bool
+}
+
+type stateParams struct {
+	instanceID         string
+	stateFilePath      string
+	quota              uint64
+	checksum           []byte
+	changeTimer        *time.Timer
+	changeTimerChannel chan bool
+}
+
+/***********************************************************************************************************************
+ * Public
+ **********************************************************************************************************************/
+
+// New creates storagestate instance.
+func New(cfg *config.Config, storage Storage) (storageState *StorageState, err error) {
+	log.Debug("Create storagestate")
+
+	storageState = &StorageState{
+		storage:             storage,
+		storageDir:          cfg.StorageDir,
+		stateDir:            cfg.StateDir,
+		statesMap:           make(map[cloudprotocol.InstanceIdent]*stateParams),
+		stateChangedChannel: make(chan StateChangedInfo, stateChannelSize),
+		newStateChannel:     make(chan cloudprotocol.NewState, stateChannelSize),
+		stateRequestChannel: make(chan cloudprotocol.StateRequest, stateChannelSize),
+	}
+
+	if err = os.MkdirAll(storageState.storageDir, 0o755); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	if err = os.MkdirAll(storageState.stateDir, 0o755); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	if storageState.watcher, err = fsnotify.NewWatcher(); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	storageMountPoint, err := fs.GetMountPoint(cfg.StorageDir)
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	stateMountPoint, err := fs.GetMountPoint(cfg.StateDir)
+	if err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	storageState.isSamePartition = storageMountPoint == stateMountPoint
+
+	go storageState.processWatcher()
+
+	return storageState, nil
+}
+
+// Close close storagestate instance.
+func (storageState *StorageState) Close() {
+	log.Debug("Close storagestate")
+
+	storageState.watcher.Close()
+}
+
+// Setup setup storagestate instance.
+func (storageState *StorageState) Setup(
+	instanceID string, params SetupParams,
+) (storagePath string, statePath string, stateChecksum []byte, err error) {
+	storageState.Lock()
+	defer storageState.Unlock()
+
+	log.WithFields(log.Fields{
+		"instaceID":    instanceID,
+		"storageQuota": params.StorageQuota,
+		"stateQuota":   params.StateQuota,
+	}).Debug("Setup storage and state")
+
+	storageStateInfo, err := storageState.storage.GetStorageStateInfoByID(instanceID)
+	if err != nil {
+		if !errors.Is(err, ErrNotExist) {
+			return "", "", nil, aoserrors.Wrap(err)
+		}
+
+		storageStateInfo = StorageStateInstanceInfo{}
+
+		if err = storageState.storage.AddStorageStateInfo(
+			instanceID, storageStateInfo); err != nil {
+			return "", "", nil, aoserrors.Wrap(err)
+		}
+	}
+
+	if storagePath, err = storageState.setup(instanceID, params); err != nil {
+		return "", "", nil, aoserrors.Wrap(err)
+	}
+
+	if _, ok := storageState.statesMap[params.InstanceIdent]; ok {
```

Comment:
`stopStateWatching` does the same check.

Reply:
Done